mypy/build.py

```diff
@@ -778,95 +785,188 @@ def is_file(path: str) -> bool:
 
-    key = (id, lib_path)
-    if key not in find_module_cache:
-        find_module_cache[key] = find()
-    return find_module_cache[key]
 
+class ImportContext:
+    """
+    Describes module import context
 
-def find_modules_recursive(module: str, lib_path: List[str]) -> List[BuildSource]:
-    module_path = find_module(module, lib_path)
-    if not module_path:
-        return []
-    result = [BuildSource(module_path, module, None)]
-    if module_path.endswith(('__init__.py', '__init__.pyi')):
-        # Subtle: this code prefers the .pyi over the .py if both
-        # exists, and also prefers packages over modules if both x/
-        # and x.py* exist.  How?  We sort the directory items, so x
-        # comes before x.py and x.pyi.  But the preference for .pyi
-        # over .py is encoded in find_module(); even though we see
-        # x.py before x.pyi, find_module() will find x.pyi first.  We
-        # use hits to avoid adding it a second time when we see x.pyi.
-        # This also avoids both x.py and x.pyi when x/ was seen first.
-        hits = set()  # type: Set[str]
-        for item in sorted(os.listdir(os.path.dirname(module_path))):
-            abs_path = os.path.join(os.path.dirname(module_path), item)
-            if os.path.isdir(abs_path) and \
-                    (os.path.isfile(os.path.join(abs_path, '__init__.py')) or
-                    os.path.isfile(os.path.join(abs_path, '__init__.pyi'))):
-                hits.add(item)
-                result += find_modules_recursive(module + '.' + item, lib_path)
-            elif item != '__init__.py' and item != '__init__.pyi' and \
-                    item.endswith(('.py', '.pyi')):
-                mod = item.split('.')[0]
-                if mod not in hits:
-                    hits.add(mod)
-                    result += find_modules_recursive(
-                        module + '.' + mod, lib_path)
-    return result
-
-
-def verify_module(id: str, path: str) -> bool:
-    """Check that all packages containing id have a __init__ file."""
-    if path.endswith(('__init__.py', '__init__.pyi')):
-        path = dirname(path)
-    for i in range(id.count('.')):
-        path = dirname(path)
-        if not any(os.path.isfile(os.path.join(path, '__init__{}'.format(extension)))
-                   for extension in PYTHON_EXTENSIONS):
+    Do we already discovered implementation?
+    What kind of module we discovered?
+    """
+    def __init__(self) -> None:
+        self.has_py = False  # type: bool
+        self.type = None  # type: Optional[ModuleType]
+        # Paths can contain only one ".py" path, but multiple stubs
+        self.paths = []  # type: List[str]
+
+    def maybe_add_path(self, path: str, type: ModuleType) -> None:
+        """
+        Add path to import context.
+        Modifies self.paths in case if arguments satisfy import context state
+        """
+        assert path.endswith((os.path.sep,) + tuple(PYTHON_EXTENSIONS))
```

Comment:
> Otherwise there is no circumstance where self.paths should ever have length > 1.

@carljm That's not true, as I understand.
Here are examples of packages split into multiple directories.
https://github.com/m-novikov/mypy-test/tree/master/split_package_stubs
Stubs and implementations located in separate directories.
You can even have split implementations:
https://github.com/m-novikov/mypy-test/tree/master/split_package_test
I am not sure if this is the desired behavior, but it is in master branch.
What if stubs are lagging in updates?

Reply:
I believe this is an accident of the current implementation. Python itself cannot find `pkg.b` in the `split_package_test` case (try `PYTHONPATH='dir1:dir2' python -c 'import pkg.b'`). Even for mypy itself, there is an inconsistency between `find_module` and `find_modules_recursive`; `find_module` can find either `pkg.a` or `pkg.b`, but `find_modules_recursive` on `pkg` will find only `pkg` and `pkg.a`, it will not find `pkg.b`. In this case `find_modules_recursive` is implementing the right Python (pre-namespace) import semantics (because of the "hits" tracking, it never accepts or uses more than one path for any module id), whereas `find_module` takes a performance shortcut (just append `pkg/` to all `lib_path` entries and see if the requested module exists there) instead of going module-by-module from the top level and verifying that we are using the correct path at each level.

In the no-stubs case, I think this can be considered purely a bug and should be fixed; mypy should not be able to find modules Python itself will not import, given the same lib path.

In the stubs case, I can imagine that some people may now be relying on this existing behavior in order to provide stubs for only some submodules in a package, while falling back to the implementation for submodules without stubs. I would leave it up to @ilevkivskyi / @gvanrossum / @JukkaL whether we need to preserve this behavior. If we do (and we do it consistently), it effectively means we need to treat packages with only `__init__.pyi` as a form of namespace package, stopping the search only when we find a package with `__init__.py`. (I don't think there's a use case for finding stub packages that occur _after_ the implementation on lib path).

Regardless, this exception should apply only for packages (ending with `__init__.pyi`); there's never any reason to track multiple paths for a regular module (we never make use of both `a.py` in one `lib_path` directory and `a.pyi` in another; one or the other always takes full priority).

Reply:
Any feedback from the core team here? Do we need to maintain the (I think accidentally implemented, but possibly useful) behavior that stub packages (with `__init__.pyi`) already in practice behave kind of like a namespace package? That is, if you have this structure:

```
.
├── src
│   └── pkg
│       ├── a.py
│       ├── b.py
│       └── __init__.py
└── stubs
    └── pkg
        ├── b.pyi
        └── __init__.pyi
```

with both `stubs` and `src` on `MYPYPATH` (in that order), mypy will still find and use `src/pkg/a.py`, even though that violates Python import's usual handling of non-namespace packages?

My feeling is that people may likely be relying on this (I think we currently are at Instagram, actually), but I don't want to put a lot of work into supporting it in this PR without confirmation that it should be formally supported, rather than fixed as a bug.

Reply:
(The same is currently also true if `stubs` contains all `.py` files rather than `.pyi` -- this I am quite sure is a bug and should be fixed.)